contracts/validator/contract/XDCValidatorV2.sol

```diff
@@ -0,0 +1,531 @@
+    }
+
+    modifier onlyCandidate(address _candidate) {
+        require(
+            validatorsState[_candidate].isCandidate,
+            "Only candidate can call this function"
+        );
+        _;
+    }
+
+    modifier onlyValidCandidate(address _candidate) {
+        require(!invalidCandidate[_candidate], "Invalid Candidate");
+        require(validatorsState[_candidate].isCandidate, "Invalid Candidate");
+        _;
+    }
+
+    modifier onlyNotCandidate(address _candidate) {
+        require(!invalidCandidate[_candidate], "Invalid Candidate");
+        require(
+            !validatorsState[_candidate].isCandidate,
+            "Already a candidate"
+        );
+        _;
+    }
+
+    modifier onlyValidVote(address _candidate, uint256 _cap) {
+        require(
+            validatorsState[_candidate].voters[msg.sender] >= _cap,
+            "Invalid Vote"
+        );
+        if (validatorsState[_candidate].owner == msg.sender) {
+            require(
+                validatorsState[_candidate].voters[msg.sender].sub(_cap) >=
+                    minCandidateCap,
+                "Minimum cap should be maintained"
+            );
+        }
+        _;
+    }
+
+    modifier onlyValidWithdraw(uint256 _blockNumber, uint _index) {
+        require(!invalidOwner[msg.sender], "Invalid Owner");
+        require(_blockNumber > 0, "Invalid block number");
+        require(
+            block.number >= _blockNumber,
+            "Block number should be less than current block number"
+        );
+        require(
+            withdrawsState[msg.sender].caps[_blockNumber] > 0,
+            "No cap to withdraw"
+        );
+        require(
+            withdrawsState[msg.sender].blockNumbers[_index] == _blockNumber,
+            "Invalid index"
+        );
+        _;
+    }
+
+    constructor(
+        address[] _candidates,
+        uint256[] _caps,
+        address _firstOwner,
+        uint256 _minCandidateCap,
+        uint256 _minVoterCap,
+        uint256 _maxValidatorNumber,
+        uint256 _candidateWithdrawDelay,
+        uint256 _voterWithdrawDelay
+    ) public {
+        minCandidateCap = _minCandidateCap;
+        minVoterCap = _minVoterCap;
+        maxValidatorNumber = _maxValidatorNumber;
+        candidateWithdrawDelay = _candidateWithdrawDelay;
+        voterWithdrawDelay = _voterWithdrawDelay;
+        candidateCount = _candidates.length;
+        owners.push(_firstOwner);
+        ownerCount++;
+        for (uint256 i = 0; i < _candidates.length; i++) {
+            candidates.push(_candidates[i]);
+            validatorsState[_candidates[i]] = ValidatorState({
+                owner: _firstOwner,
+                isCandidate: true,
+                cap: _caps[i]
+            });
+            voters[_candidates[i]].push(_firstOwner);
+            ownerToCandidate[_firstOwner].push(_candidates[i]);
+            validatorsState[_candidates[i]].voters[
+                _firstOwner
+            ] = minCandidateCap;
```

Comment:
This should be `=_caps[i]`. This is a legacy bug and we solved it on subnet. cf. https://github.com/XinFinOrg/XDC-Subnet/pull/82/files